# Import

In [1]:
import numpy as np
import pandas as pd
import os
import math

from scipy import stats #Analysis 
from scipy.stats import norm 

import missingno as msno
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import rcParams
%matplotlib inline

In [2]:
# 한글 폰트가 깨지면 실행하세요! 
# 한글 폰트가 따로 깔려있지 않은 컴퓨터에서는 해당 cell을 넘겨주세요!!! 
import matplotlib.font_manager as fm
fontlist = fm.findSystemFonts(fontpaths = None, fontext='ttf')
# 아래 주석을 지워서 폰트 리스트를 확인하고 한글 폰트를 font_path에 추가합니다
fontlist[:]

from matplotlib import font_manager, rc
#font_path = 'C:\\Users\\mtang\\AppData\\Local\\Microsoft\\Windows\\Fonts\\NanumSquare.ttf'
font_path = 'C:\\WINDOWS\\Fonts\\NanumGothicLight.ttf'
font = font_manager.FontProperties(fname=font_path).get_name()
rc('font', family=font)

# 1. Data Load  

**치매 예방을 위한 라이프로그 치매 분류**

**9,327 rows × 66 columns**

For more details https://aihub.or.kr/problem_contest/nipa-learning-platform/6

In [5]:
path = "./Dataset/"

# original data : data 
data =  pd.read_csv(path + "dementia_data.csv", parse_dates=['summary_date']) # 애초에 datatime 유형의 데이터를 읽어 올 때부터 형변환하기
data

EMAIL summary_date  activity_average_met  \
0     nia+404@rowan.kr   2020-11-27               1.71875   
1     nia+404@rowan.kr   2020-11-28               1.40625   
2     nia+404@rowan.kr   2020-11-29               1.46875   
3     nia+404@rowan.kr   2020-11-30               0.34375   
4     nia+404@rowan.kr   2020-12-01               1.46875   
...                ...          ...                   ...   
9322  nia+206@rowan.kr   2020-12-22               1.34375   
9323  nia+206@rowan.kr   2020-12-24               1.34375   
9324  nia+206@rowan.kr   2020-12-26               1.53125   
9325  nia+206@rowan.kr   2020-12-27               1.34375   
9326  nia+206@rowan.kr   2020-12-28               1.31250   

      activity_cal_active  activity_cal_total activity_class_5min  \
0                     730                2944                 ...   
1                     342                2449                 ...   
2                     401                2544                 ...   
3                      27                1850                 ...   
4                     333                2518                 ...   
...                   ...                 ...                 ...   
9322                  227                2316                 ...   
9323                  249                2351                 ...   
9324                  570                2682                 ...   
9325                  295                2331                 ...   
9326                  186                2280                 ...   

      activity_daily_movement  activity_high  activity_inactive  \
0                       14346              0                417   
1                        6352              0                473   
2                        7297              0                586   
3                         491              0                176   
4                        5861              0                646   
...                       ...            ...                ...   
9322                     3863              3                735   
9323                     4411              1                780   
9324                    11057              1                518   
9325                     5135              4                578   
9326                     3178              0                778   

      activity_inactivity_alerts  ...  sleep_temperature_deviation  \
0                              0  ...                        -0.12   
1                              0  ...                        -0.32   
2                              0  ...                         0.07   
3                              0  ...                        -0.41   
4                              0  ...                        -0.27   
...                          ...  ...                          ...   
9322                           2  ...                        -0.16   
9323                           4  ...                        -0.09   
9324                           1  ...                         0.41   
9325                           0  ...                         0.27   
9326                           1  ...                        -0.05   

      sleep_temperature_trend_deviation timezone  sleep_total  \
0                                 99.99      NaN           \r   
1                                 99.99      NaN           \r   
2                                 99.99      NaN           \r   
3                                 99.99      NaN           \r   
4                                 99.99      NaN           \r   
...                                 ...      ...          ...   
9322                              99.99      NaN           \r   
9323                              99.99      NaN           \r   
9324                              99.99      NaN           \r   
9325                              99.99      NaN           \r   
9326                              99.99      NaN           \r   

                CONVERT(acti

# 2. EDA  

**9,327 rows × 66 columns**

148명에 대한 데일리 라이프로그 데이터 
크게 걸음거리, 수면, 컨버팅 데이터로 나눌 수 있겠습니다. 

**basic (3 cols)**
- timezone 시간 장소 정보 
    - drop 
    - 완벽한 null variable 
- EMAIL
    - drop
    - 식별자 feature 
- summary_date

**activity (26 cols)**
- activity_class_5min, activity_met_1min, 하루간 5분당 활동 로그,하루간 1분당 MET 로그
    - drop 
    - BLOB type -> js로 읽을 수 있다고는 함 

**sleep (31 cols)**
- sleep_is_longest, 본 수면 여부 
    - 전부 다 1 
- sleep_temperature_trend_deviation, 피부 온도 경향 편차 
    - 전부 다 99.99
- sleep_total, 수면 시간 
    - 전부 다 \r
- sleep_temperature_deviation 
    - drop 
    - sleep_temperature_delta 랑 같은 변수 
    - sleep_temperature_deviation 을 drop 하고 delta를 남길 것 
- sleep_hr_5min, sleep_hypnogram_5min, sleep_rmssd_5min
    - 5분 당 심박동로그, 수면 상태 로그, 5분당 심박동변동 로그
    - drop 
    - BLOB type 

**converting**
- CONVERT(activity_class_5min USING utf8)    
- CONVERT(activity_met_1min USING utf8)      
- CONVERT(sleep_hr_5min USING utf8)          
- CONVERT(sleep_hypnogram_5min USING utf8)  
- CONVERT(sleep_rmssd_5min USING utf8) 

**y variable**
- DIAG_NM 
    - 정상(CN), 경도인지 장애(MCI), 치매(Dem)
    
So we may use 56 columns for training...

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9327 entries, 0 to 9326
Data columns (total 66 columns):
 #   Column                                    Non-Null Count  Dtype         
---  ------                                    --------------  -----         
 0   EMAIL                                     9327 non-null   object        
 1   summary_date                              9327 non-null   datetime64[ns]
 2   activity_average_met                      9327 non-null   float64       
 3   activity_cal_active                       9327 non-null   int64         
 4   activity_cal_total                        9327 non-null   int64         
 5   activity_class_5min                       9327 non-null   object        
 6   activity_daily_movement                   9327 non-null   int64         
 7   activity_high                             9327 non-null   int64         
 8   activity_inactive                         9327 non-null   int64         
 9   activity_inactivity_alerts    

# 3. Preprocessing 

**9,327 rows × 66 columns**

**9,327 rows × 51 columns**


**feature drop**
- timezone
- EMAIL
- sleep_temperature_trend_deviation, sleep_is_longest, sleep_total
    - 모든 값이 1, '\r', 99.99 이므로 제거 
- activity_class_5min, activity_met_1min
    - convert data
- sleep_hr_5min, sleep_hypnogram_5min, sleep_rmssd_5min
    - convert data


- CONVERT(activity_class_5min USING utf8)    
- CONVERT(activity_met_1min USING utf8)      
- CONVERT(sleep_hr_5min USING utf8)          
- CONVERT(sleep_hypnogram_5min USING utf8)  
- CONVERT(sleep_rmssd_5min USING utf8) 

**y variable**
- DIAG_NM 
    - 정상(CN), 경도인지 장애(MCI), 치매(Dem)
    
So we may use 56 columns for training...

In [7]:
# column의 수가 많기 때문에 효율적으로 관리하기 위해서 
# column의 이름들을 묶어서 진행합시다
drop_cols = ['timezone', 'EMAIL',
             'activity_class_5min', 'activity_met_1min',
             'sleep_hr_5min', 'sleep_hypnogram_5min', 'sleep_rmssd_5min',
             'sleep_is_longest', 'sleep_temperature_trend_deviation', 'sleep_total', 'sleep_temperature_deviation']
convert_cols = data.iloc[:, 60:65].columns

In [8]:
# Drop Features 
processed_data = data.drop(drop_cols, axis=1).copy()
processed_data = processed_data.drop(convert_cols, axis=1).copy()
processed_data

summary_date  activity_average_met  activity_cal_active  \
0      2020-11-27               1.71875                  730   
1      2020-11-28               1.40625                  342   
2      2020-11-29               1.46875                  401   
3      2020-11-30               0.34375                   27   
4      2020-12-01               1.46875                  333   
...           ...                   ...                  ...   
9322   2020-12-22               1.34375                  227   
9323   2020-12-24               1.34375                  249   
9324   2020-12-26               1.53125                  570   
9325   2020-12-27               1.34375                  295   
9326   2020-12-28               1.31250                  186   

      activity_cal_total  activity_daily_movement  activity_high  \
0                   2944                    14346              0   
1                   2449                     6352              0   
2                   2544                     7297              0   
3                   1850                      491              0   
4                   2518                     5861              0   
...                  ...                      ...            ...   
9322                2316                     3863              3   
9323                2351                     4411              1   
9324                2682                    11057              1   
9325                2331                     5135              4   
9326                2280                     3178              0   

      activity_inactive  activity_inactivity_alerts  activity_low  \
0                   417                           0           545   
1                   473                           0           392   
2                   586                           0           362   
3                   176                           0            34   
4                   646                           0           377   
...                 ...                         ...           ...   
9322                735                           2           155   
9323                780                           4           167   
9324                518                           1           332   
9325                578                           0           173   
9326                778                           1           138   

      activity_medium  ...  sleep_score  sleep_score_alignment  \
0                  47  ...           62                    100   
1                   8  ...           54                    100   
2                  24  ...           61                    100   
3                   1  ...           55                    100   
4                   6  ...           56                    100   
...               ...  ...          ...                    ...   
9322               17  ...           76                    100   
9323               22  ...           77                    100   
9324               61  ...           65                    100   
9325               30  ...           81                    100   
9326               18  ...           79                    100   

      sleep_score_deep  sleep_score_disturbances  sleep_score_efficiency  \
0                   41                        50                      27   
1                   49                        43                      46   
2                   56                        47                      34   
3                   35                        40                      44   
4                   64                        55                      39   
...                ...                       ...                     ...   
9322                53                        46                      74   
9323                34                        57                      76   
9324                 1                        56                      69   
9325                 9               

In [81]:
processed_data.columns

Index(['summary_date', 'activity_average_met', 'activity_cal_active',
       'activity_cal_total', 'activity_daily_movement', 'activity_high',
       'activity_inactive', 'activity_inactivity_alerts', 'activity_low',
       'activity_medium', 'activity_met_min_high', 'activity_met_min_inactive',
       'activity_met_min_low', 'activity_met_min_medium', 'activity_non_wear',
       'activity_rest', 'activity_score', 'activity_score_meet_daily_targets',
       'activity_score_move_every_hour', 'activity_score_recovery_time',
       'activity_score_stay_active', 'activity_score_training_frequency',
       'activity_score_training_volume', 'activity_steps', 'activity_total',
       'sleep_awake', 'sleep_breath_average', 'sleep_deep', 'sleep_duration',
       'sleep_efficiency', 'sleep_hr_average', 'sleep_hr_lowest',
       'sleep_light', 'sleep_midpoint_at_delta', 'sleep_midpoint_time',
       'sleep_onset_latency', 'sleep_period_id', 'sleep_rem', 'sleep_restless',
       'sleep_rmssd', 'sl